### ISSUE TITLE
Am I the only one who don't like the blending of ELF-selfplay?

### ISSUE THREAD

**Issue by billyswong** [gcp/leela-zero#1343]
Taking "shortcut" and running selfplay of converted ELF weight is rather demotivating for me on running the client. I wonder how many think like me and how many love such blending.

To me it's not about whether it will "work" as in improving LZ's strength faster, but rather the dilution of proud feeling of "Leela Zero is trained by volunteers like us distributively". @hydrogenpi certainly loves the [steroids injections](https://github.com/gcp/leela-zero/issues/1311#issuecomment-386768065) but that wording further contribute to my bad feeling that we are cheating somehow.

**Comment by hydrogenpi** [comment]
The way I see it ELF is "zero" based, LZ is "zero" based, what is the difference? "Zero" difference.

**Comment by dzhurak** [comment]
The same for me. The difference is not that big taking into account the bigger network. In a month or two we could be on the same level with our leela.

**Comment by Ishinoshita** [comment]
Felt the same in the begining and what surprised to see the main developpers jumping in fast. I was expecting more reluctance, reference to 'principles' etc...  But I totally respect the decision of the ones who made this amazing project happen through high dedication/lot of time.
Then, I think that the major achievement of such a project is not just to opensource a strong program but knowledge itself. The legacy of Deepmind not sharing all details of AG/AZ was that people has had to figured out these details, to experiment and develop a deeper understanding. In this respect, having different teams comparing their respective work (Minigo, LCZ, now FB) and trying to replicate some aspect from other teams is beneficial for better open knowledge. For example, seeing that ELF has trained from more diverse ('dirty') games came as surprise and may reshape the conception of what is the best training approach. Whether they use corporate or community ressources, computer go enthusiasts are computer go enthusiasts. And I praise their work.

**Comment by hydrogenpi** [comment]
In the past I have criticized the project for moving too slowly but there I think they are all to be praised for moving decisively and for not losing the initiative. By creating LZ-ELF and bringing it to almost performance parity with Fb ELF OpenGo and even I would argue value adding by making it multiplatform (no one has gotten ELF to compile on Windows yet) it de-incentivise others from jumping ship. Even the developer of Lizzie was openly debating whether or not to abandon further development for Leela Zero after OpenGo just came out, well because of the quick actions of the developers of LZ, this isn’t even a point of contention anymore.

**Comment by Marcin1960** [comment]
Injection of let say, 50% games from ELF pool, could speed up LZ month or two.

I am not sure, how one should deal with such temptation.

As it is now, my LeelaZeroT on KGS with  ELF net
is beating petco3 soundly. Even with 800 playouts versus 6000 !

**Comment by WYVYW** [comment]
In my opinion LZ is exploring ways to evolve. Go Seigen was not the first to use the 3-3 and hoshi in fuseki, but he made it playable and influenced all who came after him. 
Since LZ is already playing beyond our comprehension, i like to see experimenting with it and overcoming obstacles. So thanks to all LZ developers and contributers.

**Comment by dzhurak** [comment]
We don't have the games ELF used to train itself. Thus it's not an opensource project anymore if we use their network. Besides that who knows which approach is better? On how much games ELF network was trained?

**Comment by hydrogenpi** [comment]
All zero Go engines done correctly converge to the same ultimate state, its the network that is valuable not their code base. ELF cant even compile to Windows. LZ code is tons more robust and versatile

**Comment by jkiliani** [comment]
> We don't have the games ELF used to train itself. Thus it's not an opensource project anymore if we use their network. 

Opensource means open sourcecode, nothing more, so I don't see any basis for this claim. Certainly it would be nice to have their self-play database as well, but the weights seem very valuable all by themselves.

That said, I agree that finding out the best way to use data from ELF to strengthen Leela Zero is by no means simple or straightforward. It may take some experimentation in the coming weeks to figure out a good path forward now that we got AGZ20-comparable weights dropped in our lap.

**Comment by dzhurak** [comment]
> All zero Go engines done correctly converge to the same ultimate state
I am really not sure of this. The playing style of Alphazero, Galaxy, ELF is different as well as LZ. That 's one of the most interesting questions of all this learn-from-self-play thing. What playing style and what fuseki/joseki those programs will play. We are not for the strongest bot here but for the experiments and exploring.

**Comment by hydrogenpi** [comment]
Golaxy is not a zero bot. LZ has been exploring. It poineered using net2net to shortcut faster

**Comment by WYVYW** [comment]
I doubt that any human can see a difference between the playing styles of Alphazero, ELF or Golaxy.
But i must admit that one can recognize LZ fuseki, since it is very rigid. 

It was interesting to see how the infamous flower joseki has become less favoured by LZ overe the time.
And it will be just as interesting to see how LZ evolves, by utilizing different games or weights.

There will be probably always another company that has more ressources to make a stronger bot,
so it is not about making the strongest bot, but making the most versatile and open one. 

Experimenting and exploring boundaries seems therefore reasonable. 
Those who want pure blood can still use the old weights.

**Comment by hydrogenpi** [comment]
has anyone tried net2net latest 15b to 20b and average weights with elf to do a chimera to see if its stronger than current net 131

**Comment by jkiliani** [comment]
@AncalagonX apparentely tried SWA between a net2net from current networks and ELF, but it didn't work that well... not surprisingly, since networks generated by net2net are essentially untrained. I don't think there are any magic shortcuts here, just having the ELF net will help a lot. Also, we want a net that has the strength of the ELF net but is more robust in its value head if possible. There's no way around taking the time to produce a whole lot of self-play with this network...

**Comment by bastiaf** [comment]
I don’t like the shortcut too. It feels a little bit like throwing away all the computation effort of the past, and maybe the self-play games of the OpenGo network will change the playing style. And can we be absolutely sure that it is a NN that was trained from zero? (That's not supposed to be a conspiracy theory.)

Additionally using the knowledge of another project breaks the zero approach. For me it does not matter if foreign knowledge comes from human games, from other engines or from other project's networks. This is only my puristic opinion, I can understand it if for other people zero + zero keeps to be zero. :)

Nevertheless it is exciting to see progress, and it seems to me as if a little momentum has returned.

**Comment by bastiaf** [comment]
By the way the games look really weird:
http://zero.sjeng.org/view/4a9b20b9b74929a6aeee7d2514827016b16886700da641090b7907f8c89b19ce?viewer=wgo

**Comment by jkiliani** [comment]
I agree, I also commented on this and my suspected reason here: https://github.com/gcp/leela-zero/issues/1311#issuecomment-386817266

**Comment by billyswong** [comment]
@hydrogenpi I am not opposing the birth of LZ-ELF. With the 401:29 test match, it clarifies how much stronger ELF actually is compared to LZ, without advantage of CUDA-V100-specific optimization, unlike the landslide 200:0 or 198:2 match done by Facebook team. People who like to play the strongest AI can play it instantly with a lot less hurdle.

I am only opposing the training of new official LZ net with ELF selfplay games. @WYVYW said "Those who want pure blood can still use the old weights." If new official LZ nets are trained with ELF selfplay games, net#131 will be the last official "pure blood" net and always inferior. I don't consider net2net bootstrap as a non-zero shortcut since it is a reproducible procedure that can be applied to any other chess-like game AI projects, with no extra game knowledge, so it takes me by surprise that you compare ELF injection to net2net.

p.s. Their reluctance of publishing their self-play games (not even a few slices of examples like what AGZ did) is also a factor of my dislike of this blending.

**Comment by Marcin1960** [comment]
I came to the conclusion that Leela Zero is too valuable to mix it with other projects.  Let it be as it is.

**Comment by tterava** [comment]
I don't think there's any issue with adding some higher level games from the ELF nets into the pool. We are still training the LZ network so whatever strength it can draw from those higher level games is just a bonus. Also we're taking some measure so hopefully our net won't be as brittle as the ELF net with high fluctuations in winrate.

I think it's silly to talk about something like LZ having its own style and losing it by learning from higher quality games. It's almost like saying that a 10 kyu human player loses his style when he improves to 1 dan. Also as the bots become increasingly super human it's impossible for us to even talk about any kind of distinguishable styles between them.

The only noticeably peculiarity i've seen LZ do is the fuseki where it ALWAYS invades 3-3 point right away if opponent plays 4-4. However even that quirk is starting to fade with most recent networks (before ELF injection) when we increased training steps. There are situations where it opts not to invade 3-3 and does the usual keima approach instead. I didn't see anyone get worried about that.

**Comment by bastiaf** [comment]
Improving does not mean do lose an own style.

AGZ liked the early invasion at 3-3 too, and it was copied by professional players, for example KevJie played it in the first match against AGM.

**Comment by as0770** [comment]
For me it is not a Zero project anymore when knowledge from outside is used, no matter if it is human or by AI.

**Comment by roy7** [comment]
I disagree LZ is stalling. Since moving to 15 blocks, we've only had 1 promotion that took over 65K new games. The window is 500K games still I believe. When we're getting promotions from only 10% of the training window having new information, that suggests to me we're far from stalling.

**Comment by roy7** [comment]
One area LZ seems smarter is in passing. Watching this game was painful. :)

https://online-go.com/game/12759757

**Comment by as0770** [comment]
Leela is not at all stalling right now. If it is stalling earlier than AZ and ELF there is some unknown Problem to fix. If Zero is Zero, why not going on with the ELF net? This would be the most promising if we want to get stronger as fast as possible.

**Comment by hydrogenpi** [comment]
@as0770 one "real world" reason to NOT go directly with the Fb OpenGo ELF net is because (this is not a "conspiracy theory" ) say the community works on it for two months, then facebook drops a new net that is far stronger and they only had to work on it for 2 weeks... this will instantly kill community morale and spirits and WILL be the death of LZ if that were to happen. So by doing hybrid approach, it can be a repeateable process if say down the road in a few months Google says "whatever" and dumps the AGZ weights, LZ can simply once again piggyback on that. But by essentially training fb's net for fb when they could do it so much better and faster, is putting the entire community on risk of fb pulling the rug underneath it all

**Comment by emdio** [comment]
No, you're not the only one. As far as I understand it the goal of LeelaZero is to get a program as strong as possible without external help. And by external herlp I understand -at the very least- not to use but games generated by LeelaZero playing against herself.
Some say because ELF has its origin on another "zero approach" project, it doesn't break the philosophy of the project, but I disagree here. Because the same could be said about introducing human games; all in all, human knowledge in the game of go is also a "zero approach" project. We just learnt and improved with no external help.

**Comment by Marcin1960** [comment]
@hydrogenpi 

It is good there there is an impressive ELF project. And that there are others. More the better. No need to merge them all.

As Chairman Mao said: "The policy of letting a hundred flowers bloom and a hundred schools of thought contend is designed to promote the flourishing of the arts and the progress of science." :)

**Comment by tterava** [comment]
@emdio Zero doesn't mean "no external help". It means there's no human injected knowledge of Go. Also with this being an open source project, there's a chance that some percent of the games submitted are "tainted" so we can't guarantee even that 100%.

Overall the goal is to try and train as strong a network as possible without obvious Go-specific tweaks (like ladder detection) added. Also the system seems to be robust against some anomalies in the self play games so those are very unlikely to change the outcome. There's no reason to believe that ELF used human games to train it and there doesn't seem to be any non-zero tweaks in the source code.

**Comment by emdio** [comment]
@tterava Probably I should have pointed out that Zero means "no external help" to me. And -to me- once you add games -or nets- that are not the result of LeelaZero playing against itself, the spirit of the project is not the same any more. And again, to me, once you introduce external games, I don't care whether they are from humans or from another "zero alike" project. Because all in all human games can also be considered as the result of a "zero alike" project, in the sense that humans didn't receive any external help to improve their game. So if LeelaZero starts to use "external help" we could already have used it from the beginning, and would have saved some thousand hours of training a TDK program.

Probably I'd feel more comfortable with using external help once we are 99'9% sure there's no way to improve LeelaZero any more, but I wouldn't use it to save a month of training time.

**Comment by billyswong** [comment]
@roy7 Is weight#132 already "tainted"?

**Comment by Tsun-hung** [comment]
Training is totally GCP's work

**Comment by hydrogenpi** [comment]
@billyswong what makes you think 132 is tainted? There is no way it would work that fast. Most likely leanring rate is lowered (my guess). Hey if a 60% boost is "tainted", as long as it is 60% stronger, I'll take it.

**Comment by billyswong** [comment]
@hydrogenpi you said it yourself in https://github.com/gcp/leela-zero/issues/1358#issuecomment-386933231

> Okay so I'm trying to think this out logically. Facebook releases ELF OpenGo, instantly becoming by far the strongest public open source /weights Go AI program, overnight gets adopted as the baseline by many other programs.... so going forward most programs are going to be more or less the same strength if not simply identical whitelabel altogether, so what is left? Seems to be the "engine" part is now more or less solved and even I would say undifferentiatable. superhuman on a gtx 970, is basically the end of the road.

Anyone which use ELF's weight will be like "identical whitelabel". If producing strongest weight file is the sole objective of the distributed effort. *I and many others should have stopped the clients now.* Net#132 win 60% of net#131 should be meaningless as it's still weaker than ELF. Keep on the distributed effort *without* the help of ELF is a far better investment to keep the LeelaZero brand shine and stand on its own. The injection may or may not speed up LeelaZero's upgrade while we can achieve ELF level within 1-3 months while staying pure. Why the haste of "saving a few weeks time"?

**Comment by jamesho8743** [comment]
Why not just think ELF weight as some luckily randomly self-generated network? I really don't see any issue with adding some higher level games from the ELF nets into the pool. Since it has higher quality and also some weakness, the current training process can learn a lot from it without disadvantage. And, I think "Zero" doesn't mean start from complete zero or without any help. "Zero" means trying to training without "Human" bias.

**Comment by jamesho8743** [comment]
@billyswong 
"Pure blood" is just a fantasy and of little use. As you said, there will be a lot "identical whitelable" blooming. I don't think LeelaZero brand will "shine" if it lags behind to these identical whitelable. Most people want a strong bot not pure blabla bot. Does pure blood means anything or any benefit? After all, all these bots are just AGZ implementations.

**Comment by gcp** [comment]
The Facebook project is open source and ELF is about close a clone of AGZ as Leela Zero is. The places where they differ are similar to ours, i.e. areas where AGZ didn't seem optimal, and they even copied some of the fixes from us (which is why the nets were so easy to make compatible to begin with).

Not using their work is the worst kind of tribalism and NIH syndrome possible, and I will not be part of it. I will not spend a second of my time and a watt of my power bill replicating something that **is already done** and **available for all**, just so we could plaster a "Leela Zero" stamp on it, nor would I want to do the same with the clients of anyone who is contributing.

We should look at their work and see how we can use it to make something better. There seem to be opportunities there. And that's what we're trying to do now. That *is* something I'm willing to spend time on.

If you disagree, feel free to use Discord or cbaduk to set up a "clean" effort (but please not the github issues of this project...). Everything up to and including ecab and ecab's training data is free of Elves.

(You can say a lot about how big companies like FB, Google, MS do "open source" without "community" by just dropping huge NIH codebases on github and that's fair, but does not change much about the above)

**Comment by Mardak** [comment]
@billyswong the mixing in of self-play games from ELF network is not because it's just a stronger source of generating games. If you're familiar with chess, Stockfish is much stronger than Leela Chess Zero, but Stockfish games are not used to train Leela Chess Zero. Similarly, Leela (non-zero) used to be stronger than Leela Zero, but it also was not used to train Leela Zero networks. Doing something like that would truly be taking a shortcut with unknown consequences.

So this situation with using ELF network is different from Stockfish training chess or original Leela training go is that ELF network was made from an high level approach that is nearly identical to Leela Zero approach as both are just different technical implementations of Alpha Go Zero.

As gcp has stated, the ELF network is a "peek into the future" https://github.com/gcp/leela-zero/issues/1311#issuecomment-386563423 as if we just continued down normal self-playing, training, promoting with our distributed community, we would end up with a network probably not too different. For example, looking at josekis, Leela Zero networks and ELF network do seem to match up with those published with Alpha Go Zero.

To flip things around, is it fair for the community to waste their time and electricity redoing something to end up at a similar state as the current ELF network? Or would it be better for the community to be making new progress and discoveries?

**Comment by yufanyufan** [comment]
Can we prove that all zero approach will converge to the same policy or play style?
Is it possible for leela zero to converge to very a different play style but equally strong as ELF?
